content/blog/2019-09-25-type_checker.md

```diff
@@ -0,0 +1,100 @@
+3. Keeping track of the line number and returning proper error message when encountered with an type checking error. To track the line number, the type checker works on the Bril text representation and is developed with its own parser. When scanning through the code line by line after the first pass, a line number is maintained during type checking to help the programmer debug. In order to maintain the line number with blank lines and comments, we take in the original code snippet and ignore these lines while updating it during type checking.
+
+
+
+## Evaluation and results
+
+We wrote a set of benchmarking programs to test our implementation for various types of rules defined in the design section above.
+The general test cases are classified into two sub-directories: [should-fail](https://github.com/tissue3/bril/tree/master/test/type-check/should-fail) and [should-pass](https://github.com/tissue3/bril/tree/master/test/type-check/should-pass). It has expected output named as `*.out` corresponding to the input file `*.bril`. User can simply run all test cases in a directory by running `turnt directory/*.bril`. For example, to run should-fail benchmark, one can just run `turnt test/type-check/should-pass/*.bril` at the main directory of bril.
+Because should-pass cases are trivial, where we enumerate list existing operations, in the following table we only list tests cases that our type checker will report an error message. The second column of the table aims to help one understand why an error would be reported by our type checker.
+
+| Instruction | Type Checking Rule |Testing Code Snippet  |
+| ----------- | :-----------: | :-------------: |
+| Conflict Definition  |   Variables cannot be redefined with a different type.   |      v: int = 5;<br/> v :bool = true;      |
+| Arithmetic         |          Adding an integer and a boolean type.           |                a: int = 4;<br/>b: bool = true;<br/>c: int = add a b;               |
+| Boolean         |      Cannot assign the output of boolean to an integer.      |    b: bool = true;<br/> a:int = not b;    |
+| Const       |    Cannot assign an integer const to a bool variable.    |            a: int =  const true;             |
+| Cond Branch |    Only takes bool variable as input (and 2 labels).     |     a: int = 1;<br/> br a here there;<br/> here: print a;<br/>there: jmp here; |
+
+Finally, though not required, we also implemented other checking passes as long as an input stream is parsable by bril2json. For example, `v0: boolean =const true;` is a legal statement for bril2json, but boolean is not an existing type in bril, so our type checker will report the error. Such kind of checking pass includes type existence checking, number checking, argument existence checking, label existence checking and label repentance checking.
```

Comment:
Again, no need to fix this now, but I don't quite understand why there are two categories of tests. As far as I can tell from your implementation, all of these depend on being parseable with my Lark parser ("bril2json"), so it doesn't seem right to say that these ones are the ones that are relevant in that case.

Reply:
There are two categories because I am trying to make distinction between type check and anything should be checked. For example, I feel label check has nothing to do with type, but it's parsable and an error should be reported.

Reply:
Sounds reasonable! I was confused by the reference to bril2json, but that distinction makes sense to me. :)